### PR TITLE
1543 emptry crl file fails http tpc pull transfer

### LINKED ISSUE #1543: [XrootdHttpTPC] Empty CRL file fails a HTTP TPC pull transfer
Hello,

After the submission of an HTTP TPC Pull transfer, I got the following error client side:

```
* Connection #0 to host xrootd-ccaffy-dev01.cern.ch left intact
failure: Failed to load CRL file (path? access rights?, format?)[xrootddev@xrootd-ccaffy-dev01 xrootd-run]$ 
```
On server side:

```
211027 15:06:03 32354 TPC_PullRequest: event=TRANSFER_FAIL, local=/tmp/tpcfile1, remote=https://eospps.cern.ch/eos/user/ccaffy/file1, user=(anonymous), bytes_transferred=0; HTTP library failure: Failed to load CRL file (path? access rights?, format?)
```
Here is the configuration file I have:

```
all.export /tmp nolock
all.adminpath  /var/spool/xrootd
xrd.port 1095
ofs.tpc autorm ttl 7 15 xfr 9 pgm /home/xrootddev/xrootd-run/my_xrdcp.sh
xrd.tls /etc/grid-security/xrd/xrdcert.pem /etc/grid-security/xrd/xrdkey.pem
xrd.tlsca certdir /etc/grid-security/certificates
http.exthandler xrdmacaroons libXrdMacaroons-5.so
macaroons.secretkey /etc/xrootd/macaroon-secret
ofs.authlib libXrdMacaroons-5.so
if exec xrootd
  xrd.protocol http:1095 libXrdHttp-5.so
  http.exthandler xrdtpc libXrdHttpTPC-5.so
  http.secxtractor libXrdVoms-5.so
fi
all.sitename xrootddev
```

After debugging the xrootd process, I could see that curl complains with "Failed to load CRL file (path? access rights?, format?)"
when the `CURLOPT_CRLFILE` option points to an empty file.

The workaround I found was to set the following environment variable:

```
export XRDTPC_CADIR=/etc/grid-security/certificates/
```
This prevents the class  `XrdTlsTempCA` from being instanciated and therefore prevents to
set the `CURLOPT_CRLFILE` option (done by the method `TPCHandler::ConfigureCurlCA()`).

In my opinion, this can be problematic for users. The directory where the certificates
are located is passed via the configuration of the server. If a user has no CRL file on the certificate directory, XRootD should
just ignore it and should not try to set the `CURLOPT_CRLFILE` curl option.

In production, everything works fine because the concatenated CRL file is not empty:

```
[root@eospps-fe1 (mgm:master mq:master) ~]$ cat /tmp/mgm/.xrdtls/crl_file.pem | wc -l
12518
```

My first question is, is it me who wrongly configured the server?

Otherwise, should we add a check that verifies that the concatenated CRL file is not empty before assigning it to `CURLOPT_CRLFILE` ?

@abh3 , @bbockelm what is your opinion about this issue?

Thanks in advance for your answers :)

Cheers,
Cedric

### PR DESCRIPTION
**Pull request by ccaffy**
Fixes #1543

### CODE REVIEW

**Comment by bbockelm** [comment]
Does this actually fix #1543 though?  In the last comment there, you mention it fails if the CRL file is non-empty and contains no valid CRLs.  Shouldn't we use some of the internal XRootD routines to parse the file and see if there's one CRLs present?

**Comment by abh3** [comment]
Yes, this is something to conisder. However, we should also make sure that 
malicious injection of an empty/invalid CRL file does not completely 
eliminate CRL checking for all time. This is a big security loophole. As 
much as I dislike CRL's they do serve a purpose. So, the solution is not 
completely simple.

**Comment by bbockelm** [comment]
> However, we should also make sure that malicious injection of an empty/invalid CRL file does not completely eliminate CRL checking for all time

This is a file periodically generated by XRootD in the admin directory.  I figure if the attacker owns the admin directory, then it's already game over.

It's also an aggregate of all CRLs available - if the attacker controls all your CRLs, then it's probably also game over.

**Comment by bbockelm** [comment]
Indeed, that's a fantastic idea!  I expect this to be a pretty rare occurrence so a warning message seems warranted.

@ccaffy - does this sound good to you?

**Comment by ccaffy** [comment]
Hi all, thanks for your comments!

To sum things up, you would like to disable the CRL checking and display a warning message if there is at least one wrong CRL encountered during the CRL processing?

That's sounds a good plan to me :)

**Comment by ccaffy** [comment]
> Does this actually fix #1543 though? In the last comment there, you mention it fails if the CRL file is non-empty and contains no valid CRLs. Shouldn't we use some of the internal XRootD routines to parse the file and see if there's one CRLs present?

I thought that checking that the aggregated-CRL file is not empty would be sufficient as curl already checks the validity of its content. In my opinion, failing a transfer due to an empty CRL file is not acceptable, but failing a transfer due to a wrong content is acceptable.
Hence why I only added a check on the emptiness of the file ;)

**Comment by bbockelm** [comment]
> To sum things up, you would like to disable the CRL checking and display a warning message if there is at least one wrong CRL encountered during the CRL processing?

Not quite -- I'm thinking that if there are *no* CRLs in the CRL file then we disable CRL checking.  Here, I'm willing to define "valid" as simply parseable as a CRL (not looking closely at the contents).  Particularly, we should have a routine like [this one](https://github.com/xrootd/xrootd/blob/master/src/XrdTls/XrdTlsTempCA.cc#L169-L203); instead of looping through all CRLs in a file, return successful as soon as a single one is present.

Actually, that suggests a simpler approach.  Since we are writing the CRLs to this file in the first place, why not keep track of the number written and, if there were zero, simply don't invoke the corresponding curl option?

**Comment by ccaffy** [comment]
> Actually, that suggests a simpler approach. Since we are writing the CRLs to this file in the first place, why not keep track of the number written and, if there were zero, simply don't invoke the corresponding curl option?

Yes, for me this is fine :) I can implement that :+1:

**Comment by ccaffy** [comment]
I updated the code yesterday :) Please tell me if this is OK for you so that this fix can be merged :)

File: src/XrdTls/XrdTlsTempCA.cc [hunk 1]
```diff
@@ -198,11 +209,17 @@ CRLSet::processFile(file_smart_ptr &fp, const std::string &fname)
             return false;
         }
     }
+    if(!m_atLeastOneValidCRLFound)
+        m_atLeastOneValidCRLFound = atLeastOneValidCRLFound;
```
Comment (abh3) [inline src/XrdTls/XrdTlsTempCA.cc:213]: Doesn't this mean that once a valid CRL is found a subsequent scan that doesn't find one will not be registered as such? Won't that be a problem?

**Review by abh3 — COMMENTED** [review]
A couple of questions and perhaps needed changes before we merge this. Sorry for taking so long.

File: .gitignore [hunk 2]
```diff
@@ -58,3 +58,5 @@ xrootd.spec
 dist
 *.egg-info
 bindings/python/VERSION
+cmake-*
+.idea/
```
[the remark this answers is not in the document]
Comment (ccaffy) [reply .gitignore:62]: Hi Andy,

No problem :)

Because I use CLion as IDE with local code deployed on a VM. The cmake-* directories is the equivalent of the `build` directory.
The .idea directory is created by CLion.
I just do not want these directories to be pushed to the project :)

Now that you said this, it looks like I can put these excludes in my local `.git/info/exclude` file. I will therefore remove these two lines from the .gitignore file

File: src/XrdTls/XrdTlsTempCA.cc (hunk 1 above)
Comment (ccaffy) [reply src/XrdTls/XrdTlsTempCA.cc:213]: I only modify the variable `m_atLeastOneValidCRLFound` if this variable is false. So, once a valid CRL is found, the `m_atLeastOneValidCRLFound` will be set to true and therefore will never be assigned anymore during the `XrdTlsTempCA::Maintenance()` loop.

That way, we only enable CRL checking if there is at least one valid CRL found during the Maintenance run :)

Comment (abh3) [reply src/XrdTls/XrdTlsTempCA.cc:213]: What about this little issue that doesn't seem quite right.

Comment (abh3) [reply src/XrdTls/XrdTlsTempCA.cc:213]: Yes, but you should consider that the next time the maintenance loop runs there may be no valid CRLs. I know that's unlikely but unlikely things happen. I think you should always update the current state of the maintenance loop otherwise that look does half the j0ob.

Comment (ccaffy) [reply src/XrdTls/XrdTlsTempCA.cc:213]: If during the next time the maintenance loop runs there is no valid CRL file, do we agree that the CRL checking must be disabled?

Comment (ccaffy) [reply src/XrdTls/XrdTlsTempCA.cc:213]: Ok, I see what you mean. I just need to reset this boolean to false at the beginning of the Maintenance run :+1:

Comment (abh3) [reply src/XrdTls/XrdTlsTempCA.cc:213]: I think that if the current CRL file is replaced with a bad CRL file then, yes. It should be disabled. This is dynamic because the maintenance check tries to create a CRL file and if it fails then a) if the previous CRL file is still enforce then we need do nothing, or b) it gets replaced with a null file then we need top disable CRL checking. So, in the end it depends on the state the maintenance run leaves the CRL file (something I have not researched).

**Comment by abh3** [comment]
Well, not really because it is set false at the start (well initial 
state). Let me explain the state:

a) we find a CRL, the state is set to true.

b) second scan and we don't find a CRL.

c) What do we do? Two cases:
    1) The previous file is kept so it's OK to say use the CRL.
    2) The previpus file is replaced with a null file -- not OK.

I think you see the conundrum. It really depends on what happens to the 
compisite CRL file; i.e. is it kept or replaced. So, given that I don't 
know (as a reviewer) it seems safest to just reflect the current state. 
Now, you can elucidete what the case is. However it should be made 
painfully clear in the comments ahead of the test; otherwise we will be 
revisting this for years to come.

Andy

**Comment by ccaffy** [comment]
I thought we wanted the composite CRL file to reflect what is in the `/etc/grid-security/certificates` directory and not trying to see if in the previous run, a CRL that was found is not found anymore or is replaced by a null file in the next run.

Here are the tests I did:

First, best case scenario, there are two valid pem files in the `/etc/grid-security/certificates` directory:

```
//Server side
211201 13:42:09 23886 TPC_TempCA: Reloading the list of CAs and CRLs in directory
```

Composite CRL file:

```
-----BEGIN X509 CRL-----
[...] CRL1 file content
-----END X509 CRL-----
-----BEGIN X509 CRL-----
[...] CRL2 file content
-----END X509 CRL-----
```

Now, I empty the CRL2 file content:
```
root@xrootd-ccaffy-dev01 xrootd-run]# cat /etc/grid-security/certificates/my_crl_file_2.crl
[root@xrootd-ccaffy-dev01 xrootd-run]#
```

... and after the next run of the Maintenance:

```
-----BEGIN X509 CRL-----
[...] CRL1 file content
-----END X509 CRL-----
```

The composite CRL file is re-generated. Containing only the first CRL file as I emptied the second CRL file.

Now I will modify the CRL1 file content to break the CRL correctness of this file:

```
[root@xrootd-ccaffy-dev01 xrootd-run]# truncate -s 500 /etc/grid-security/certificates/my_crl_file.crl
```

Then the composite CRL file is re-generated and contains nothing:

```
[xrootddev@xrootd-ccaffy-dev01 xrootd-run]$ cat /var/spool/xrootd/.xrdtls/crl_file.pem
[xrootddev@xrootd-ccaffy-dev01 xrootd-run]$
```

A TPC transfer will then have the CRL check disabled:

```
211201 13:56:48 23872 TPC_TpcHandler: No valid CRL file has been found in the file /var/spool/xrootd/.xrdtls/crl_file.pem. Disabling CRL checking.
```

Is this the behaviour we expect? Because for me, the composite CRL file should be an image of what is in the `/etc/grid-security/certificates` directory.

To sum things up, the composite CRL file is completely re-generated at each Maintenance run :)

Sorry for the time spent on this topic, but I really need to understand and I want to be sure that you have the full picture of why I did this implementation like this ;)

**Comment by xrootd-dev** [comment]
I don't think that would work all that well because there may be a 
transfer during the maintenance run and then CRLs would not be checked 
eventhough there still is the previous file. Franklly, I would just get 
rid of the if statement gaurd.

Andy

**Comment by ccaffy** [comment]
Alright then, I get rid of the if statement. Thanks for your comments :)
